AgentDiscoveries-Backend/src/main/java/org/softwire/training/api/routes/v1/AgentsRoutes.java

```diff
@@ -34,13 +35,27 @@ public Agent createAgent(Request req, Response res) throws FailedRequestExceptio
     }
 
     public Agent readAgent(Request req, Response res, int id) throws FailedRequestException {
+        int userId = req.attribute("user_id");
+        if(id!=userId){
+            throw new FailedRequestException(ErrorCode.INVALID_CREDENTIALS, "User not authorised");
+        }
         return agentsDao.getAgentByUserId(id)
                 .orElseThrow(() -> new FailedRequestException(ErrorCode.NOT_FOUND, "Agent not found"));
     }
 
     public Agent updateAgent(Request req, Response res, int id) throws FailedRequestException {
+        int userId = req.attribute("user_id");
+        if(id!=userId){
+            throw new FailedRequestException(ErrorCode.INVALID_CREDENTIALS, "User not authorised");
+        }
+
         Agent agent = JsonRequestUtils.readBodyAsType(req, Agent.class);
-        agent.setUserId(id);
+        Optional<Agent> optionalAgent = agentsDao.getAgentByUserId(userId);
+        if (!optionalAgent.isPresent()) {
+            throw new FailedRequestException(ErrorCode.NOT_FOUND, "user Id not found");
+        }
+        Agent oldAgent = optionalAgent.get();
+        agent = mergeAgents(agent, oldAgent);

```

Comment:
Why do we need to merge? Will the update request not include all the parameters?

Reply:
Not necessarily; for example the front end currently doesn't necessarily know the agents name/d.o.b

Reply:
Alright. Doing fetch/merge/update does introduce a race-condition with multiple updates, but not too much to worry about in our little app.

(Otherwise we could update only the fields which are set - though then you may need to distinguish between 'not specified' and explicitly null)